File: Tutorials/LeetCode/Solutions/29-Divide_Two_Integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        INT_MAX = 2**31 - 1
        INT_MIN = -2**31
        
        # Handle overflow case
        if dividend == INT_MIN and divisor == -1:
            return INT_MAX
        
        # Determine sign
        negative = (dividend < 0) != (divisor < 0)
        
        # Work with absolute values
        dividend = abs(dividend)
        divisor = abs(divisor)
        quotient = 0
        
        # Exponential search with bit shifting
        while dividend >= divisor:
            temp_divisor, multiple = divisor, 1
            
            # Double until we exceed dividend
            while dividend >= (temp_divisor << 1):
                temp_divisor <<= 1
                multiple <<= 1
            
            # Subtract and accumulate
            dividend -= temp_divisor
            quotient += multiple
        
        # Apply sign and return
        return -quotient if negative else quotient

File: Tutorials/LeetCode/Solutions/29-Divide_Two_Integers.rs
impl Solution {
    pub fn divide(dividend: i32, divisor: i32) -> i32 {
        if dividend == i32::MIN && divisor == -1 {
            return i32::MAX;
        }
        
        let negative = (dividend < 0) ^ (divisor < 0);
        let mut dvd = (dividend as i64).abs();
        let dvs = (divisor as i64).abs();
        let mut result = 0;
        
        while dvd >= dvs {
            let mut shift = 0;
            while dvd >= (dvs << (shift + 1)) {
                shift += 1;
            }
            result += 1 << shift;
            dvd -= dvs << shift;
        }
        
        if negative { -result as i32 } else { result as i32 }
    }
}